added json output support for stats handler

ramaraochavali opened the pull request:
PR for [issue](https://github.com/envoyproxy/envoy/issues/1862)

ramaraochavali commented on source/server/http/admin.cc:
```diff
@@ -279,11 +289,42 @@ Http::Code AdminImpl::handlerStats(const std::string&, Buffer::Instance& respons
+      response.add(fmt::format("{}: {}\n", stat.first, stat.second));
+    }
+  } else {
+    std::string format = params.begin()->first;
```
Tried to use Field, FieldPtr and create json but not able to include them here.

mattklein123 replied:
Yeah, thanks for trying. See TODO request above.

mattklein123 commented:
@ramaraochavali before I dive into full review, a few things:
1) Will need DCO (see contributing guide).
2) Please get the PR compiling/passing tests.
3) Will need docs here: https://envoyproxy.github.io/envoy/operations/admin.html
4) Will need a test here https://github.com/envoyproxy/envoy/blob/master/test/integration/integration_admin_test.cc that the body returned is actually valid JSON.

ramaraochavali commented:
@mattklein123  I have updated the doc and added an integration test that validates the json. I have signed the DCO in my last commit. But still it shows as failed? Do I have to commit with a new branch?
All tests are passing. Can you please review now?

mattklein123 commented:
@ramaraochavali every commit has to have DCO. Please just squash/rebase and force push the entire PR on current master (to fix OSX tests) and then we can review. Thank you.

ramaraochavali commented:
@mattklein123 done..did a force commit as single one

mattklein123 commented on docs/operations/admin.rst:
```diff
@@ -137,3 +137,7 @@ The fields are:
+
+  .. http:get:: /stats?format=json
+
+    Same as /stats but outputs them in json format. This can be used for programmatic access of stats.
```
nit: "Same as /stats the output is in JSON format"

mattklein123 commented on source/server/http/admin.cc:
```diff
@@ -31,8 +31,17 @@
 #include "common/upstream/host_utility.h"
 
 #include "fmt/format.h"
+#include "rapidjson/document.h"
```
I would prefer that we don't like rapidjson out into the rest of the code, but it's OK for now. Can you please add above the includes:

"TODO(mattklein123): Switch to JSON interface methods and remove rapidjson dependency"

mattklein123 commented on source/server/http/admin.cc:
```diff
@@ -279,11 +290,42 @@ Http::Code AdminImpl::handlerStats(const std::string&, Buffer::Instance& respons
-  for (auto stat : all_stats) {
-    response.add(fmt::format("{}: {}\n", stat.first, stat.second));
+  if (params.size() == 0) {
+    // No Arguments so use the standard
```
nit: "No arguments so use the standard output." (period end of sentence)

mattklein123 commented on source/server/http/admin.cc:
```diff
@@ -279,11 +290,42 @@ Http::Code AdminImpl::handlerStats(const std::string&, Buffer::Instance& respons
+    }
+  } else {
+    std::string format = params.begin()->first;
+    std::string formatvalue = params.begin()->second;
```
Please use snake case for all vars (e.g., "format_value", "stats_array" below, etc.). Won't comment on each one.

ramaraochavali replied:
will change.. this habit from my lot of java coding :-)

mattklein123 commented on source/server/http/admin.cc:
```diff
@@ -279,11 +290,42 @@ Http::Code AdminImpl::handlerStats(const std::string&, Buffer::Instance& respons
+    std::string format = params.begin()->first;
+    std::string formatvalue = params.begin()->second;
+    if (format == "format" && formatvalue == "json") {
+      rapidjson::Document document;
```
nit: Can we move the actual JSON creation into a helper function. We will likely have other output types in the future.

ramaraochavali replied:
will do

ramaraochavali commented:
@mattklein123 Thanks for your quick review. Addressed all the comments. Should be good to go. Let me know if further changes are required

mattklein123 commented on docs/operations/admin.rst:
```diff
@@ -137,3 +137,7 @@ The fields are:
+
+  .. http:get:: /stats?format=json
+
+  Same as /stats the output is in JSON format. This can be used for programmatic access of stats.
```
Please switch first sentence to "Outputs /stats in JSON format."

mattklein123 commented on source/server/http/admin.cc:
```diff
@@ -267,9 +278,11 @@ Http::Code AdminImpl::handlerServerInfo(const std::string&, Buffer::Instance& re
   // We currently don't support timers locally (only via statsd) so just group all the counters
   // and gauges together, alpha sort them, and spit them out.
+  Http::Code rc = Http::Code::OK;
+  Http::Utility::QueryParams params = Http::Utility::parseQueryString(url);
```
nit: const

mattklein123 commented on source/server/http/admin.cc:
```diff
@@ -279,11 +292,47 @@ Http::Code AdminImpl::handlerStats(const std::string&, Buffer::Instance& respons
+      response.add(fmt::format("{}: {}\n", stat.first, stat.second));
+    }
+  } else {
+    std::string format_key = params.begin()->first;
```
nit: const for this var and the the following

mattklein123 commented on source/server/http/admin.cc:
```diff
@@ -279,11 +292,47 @@ Http::Code AdminImpl::handlerStats(const std::string&, Buffer::Instance& respons
+    std::string format_key = params.begin()->first;
+    std::string format_value = params.begin()->second;
+    if (format_key == "format" && format_value == "json") {
+      std::string json_stats = statsAsJson(all_stats);
```
nit: don't need local var, just fold function call into the following line.

mattklein123 commented on source/server/http/admin.h:
```diff
@@ -115,7 +115,7 @@ class AdminImpl : public Admin,
                       const Upstream::Outlier::Detector* outlier_detector,
                       Buffer::Instance& response);
-
+  std::string statsAsJson(std::map<std::string, uint64_t> all_stats);
```
nit: newline after this line.

ramaraochavali commented:
@mattklein123 I think have addressed all comments. Have a final look once. Sorry because of lack of my recent c++ coding exp, took more time for you to review I guess. :-)

mattklein123 commented on source/server/http/admin.h:
```diff
@@ -115,6 +115,7 @@ class AdminImpl : public Admin,
   void addOutlierInfo(const std::string& cluster_name,
                       const Upstream::Outlier::Detector* outlier_detector,
                       Buffer::Instance& response);
+  std::string statsAsJson(std::map<std::string, uint64_t> all_stats);
```
oops sorry one more thing, can you make this `const std::map<std::string, uint64_t>&`. Sorry! :)

ramaraochavali replied:
No issues. committed..please check
